### ISSUE TITLE
Sprint Poker ‘Estimate’ phase AC

### ISSUE THREAD

**Issue by ackernaut** [ParabolInc/parabol#4092]
This is the output of #4017 

### Acceptance criteria

- [x] Let’s review the following, decide where we need consent or iteration, and then break down the updated AC bullets into their smaller tasks where it makes sense
- Effort: 1 hour
- [See Figma](https://www.figma.com/file/csERjrh67KZKgnT5VahFSb/Sprint-Poker?node-id=941%3A91)

#### 1 stage + 1 or many dimensions
- The phase has 1 stage per story
- Each stage can have 1 or many dimensions. For v1 we only allow a single dimension in the UI.
- Each dimension has 2 states: [voting](https://www.figma.com/file/csERjrh67KZKgnT5VahFSb/Sprint-Poker?node-id=941%3A91) or [discussion](https://www.figma.com/file/csERjrh67KZKgnT5VahFSb/Sprint-Poker?node-id=939%3A8570)
- Each dimension starts in the voting state
- When all teammates have voted for a dimension, the dimension switches to the discussion state automatically
- The facilitator can manually reveal, switching to the discussion state
- A single thread in the sidebar is available to discuss during any point in the stage across 1 or many dimensions. For v1 on mobile viewports, we’ll look at toggling the thread as a sidebar since there’s not room for both a thread and the main stage view. For laptop+ it’ll probably always be expanded until we add that pattern (thread as a sidebar).
- Users can navigate to any stage or dimension at any time to allow for async-first participation
- When the facilitator is connected, the users will be tethered to both the stage (story) and the dimension the facilitator is viewing when there are many. Hitting `Rejoin` takes folks to the dimension everybody is assumed to be discussing because of the facilitator’s location.

#### Story task card
- Has a title (first line without breaking)
- Optionally has a second line as the beginning of the description or first comment from the integrated issue
- Both lines truncate with ellipsis to keep the card height to a minimal to save room for the components for voting and discussing the results
- On mobile viewports, tapping the expand icon opens up a ‘details’ view as a dialog. On laptop+ viewports, it expands inline pushing the voter avatar grid down in the overall scrollable content area. The cards and bottom bar remain pinned and over any content that gets pushed down.
- If the task card is integrated, there’s a link to the external issue
- If the task card is not integrated, there’s an icon to push to an external service via integration
- If the task card is not integrated, the content can be edited
- If the card can be edited, it’s edited in the expanded or dialog view. Pressing shift will break from the title line to the description line.

### The voting state (dimension column)

- [As folks vote](https://www.figma.com/file/csERjrh67KZKgnT5VahFSb/Sprint-Poker?node-id=941%3A91), their avatars fly in from the right (off-screen) next to a card with a check
- When I haven’t voted yet there’s a prompt ‘Tap a card to vote, swipe to see other dimensions’
- When I vote, my avatar flies into the group as well
- The card that I used for that dimension will animate up to the selection state when I navigate between dimensions
- When the row of avatars gets long we’ll add the overflow badge (see note below). The avatar group can be tapped to see a read-only list of everyone who’s voted in a dialog. Mobile ~7 avatars or 6 + badge, Laptop+ ~12 or 11 + badge
- When there’s an overflow badge, we consider animating the change in the count instead of animating an avatar into the group.
- When all votes are in it automatically switches to the discussion state
- The facilitator can manually reveal votes for a dimension
- When there are no votes, the facilitator sees a tip. When there are 2 votes the facilitator sees a reveal button close to the avatar group.

### The discussion state (dimension column)
- [See Figma](https://www.figma.com/file/csERjrh67KZKgnT5VahFSb/Sprint-Poker?node-id=939%3A8570)
- All the avatars animate to their respective value group once the votes are revealed
- If a single card/value has many avatars on its row, we will have an overflow for the mobile column width and the laptop+ column width. It’s probably something like: on mobile, show 7 avatars or 6 avatars with a count badge +n and on laptop+ show 12 avatars or 11 with a count badge. Tapping the avatar group when there’s overflow will open a read-only list in a dialog so one can see everybody who voted.
- The mini cards representing values form a left edge in the column
- We sort from `?` cards to the highest values descending to the lowest values, followed by `Pass` cards
- The group that has the most consensus gets highlighted in some way
- If folks change their vote while in discussion mode, their avatars animate to their new group in real-time
- The facilitator has an input to add the final value. It looks like an empty card at the top of the list, like the other cards. It represents what the team is going to agree to. This input is autofocused. This is the primary action of this state so it needs the right amount of emphasis.
- When they enter the value the left nav updates. The stage has a value if there’s a single dimension. The total points for this session updates. (If we add an enhanced template later for weighted values, we could put the weighted value in the left nav).
- The facilitator has a secondary control to call a revote for this dimension. It completely clears the votes to make it obvious when people add their new vote (their avatars animate in).

#### Hand of cards
- There’s 1 card for each value in the scale plus the ‘?’ and ‘Pass’ cards. Note: the ‘?’ and ‘Pass’ cards can probably be removed if a team clones one of our public templates and updates the scale.
- Hovering a card gives feedback that it’s clickable
- Clicking or tapping a card raises the card and sets the active viewer’s vote for the current dimension
- If they already voted but were changing, the previous card animates back down and the new card animates back up
- When a dimension is in the discussion state, and there’s a single dimension, the hand of cards animates out to the right, leaving the card with the user’s vote in the bottom-right corner, like the footprint of a floating action button. Tapping the card animates the full hand out with the current vote card raised.
- When there are multiple dimensions in the future, swiping through will raise and lower any cards that have been used to vote e.g. I voted `5` on value and `8` on effort. As I swipe the 5 lowers and the 8 raises. The cards stay present instead of animating to the side to form a deck in the discussion state until we validate a better interaction and animation.

#### Bottom bar
- Both the facilitator and participants get a Next control to allow for async-first navigation that goes to the next story or dimension, whichever comes first. The ‘Next’ control is more about navigation than processing the current item. We want to anchor the facilitator to reveal and add the final value as actions in the dimension column, not the meeting’s bottom bar which are for more general purposes. 
- We exclude a Ready button for v1 and learn. The UX is centered around voting and discussion. They will see who has voted, indicating readiness to discuss. The discussion has a goal, to enter a final value. It’s a clear point to move on.
- ~Participants get a Ready control that’s scoped per dimension to indicate they’re ready to move on to the next story or dimension, whichever comes first.~
- ~We scope the Ready button per dimension. We disable it during voting and enable it during the discussion. The thinking is that we decouple readiness from voting. We’ll see voting activity in the column (more avatars!). Readiness is about indicating we need to move on as a team aka “I’m done discussing”.~

#### Future target considerations
- We allow for many dimensions per stage
- We offer swipeable columns to navigate through the dimensions
- The UI focuses on a single column at a time
- The voting or discussion state is set per dimension. One dimension can be revealed while the next one is still in the voting state.
- There’s a single thread for all dimensions i.e. 1 thread for 1 stage representing a story
- The video call component is added to the thread sidebar in the future
- We add weighted formula templates
- There’s a way to view a table of all dimensions per story instead of 1 dimension column at a time, especially for teams using a weighted formula template

**Comment by ackernaut** [comment]
### Updates to merge with above
- Users will be tethered to the facilitator when the facilitator is connected. The Rejoin action will navigate to both the story stage and the dimension column where the facilitator is (when there are multiple dimensions).
- We’ll sort from ? to highest value to lowest value when revealing votes. A card will be above the list representing the input value needed from the team, which the facilitator adds and others see read-only
- The user’s avatar will be added to the group of avatars that voted
- We need to look at the layout with the default color and letter avatars to see how it may clash with the card design
- We’ll only show the mode by the group, not in 2 places
- The reveal action will be close to the avatar group of folks who’ve voted. It only shows when there have been two or more votes to draw attention to it
- We’ll work on tucking the thread into a sidebar for mobile to start moving toward a right sidebar for thread and video
- When a template has a single dimension we’ll animate the cards over to the right when in discussion mode
### Next steps
- I’ll update the AC above, once updated we can start pulling the AC into their own issues as implementation is done
- I’ll update Figma views to reflect the latest details and show various states now that we’ve decided, to guide implementation

**Comment by jordanh** [comment]
This is epic, y'all

**Comment by mattkrick** [comment]
i think we've captured all of this in smaller issues, will close once the phase is complete

**Comment by ackernaut** [comment]
@mattkrick were the behaviors of the bottom bar captured in the general issue for implementing the meeting type?

**Comment by mattkrick** [comment]
yep! the way the design shook out, 1 dimension is its own stage, so the next button will work as intended without any extra code.
removing the isReady button is easy...
...or maybe that's where we show how many still need to vote? 🤔

**Comment by ackernaut** [comment]
We could use the circle for a percentage of folks who’ve viewed at some point. We could use the vote for the implicit ready. If folks are moving their mouse over their hand and they’ve already selected a card, we could wiggle their avatar to indicate that maybe they’re not settled yet.

At any rate, I’d like the participants to be able to move forward independently with their own next button.

We have room for experimentation for the participants’ bottom bar as we could have up to four: tips / next / ready / rejoin — I really don’t think we need a ready control

**Comment by ackernaut** [comment]
### Burndown
Most of the specs above are accounted for somewhere in an implementation issue. Note that the newer implementation issues may include changes to the specs above that reflect the latest thinking around the smaller details.

Some things I’m checking for:
- [ ] The thread is wired up, will be included in #4317
- [ ] The voting UI is in place, has been started in #4320
- [ ] The bottom bar allows participants to go to the ‘Next’ story or dimension independent of the facilitator (see the **Bottom Bar** section above)
- [ ] The ready button goes away (see the **Bottom Bar** section above)
- [ ] Tethering is scoped to each dimension. Participants can rejoin when not on the same dimension (this work is probably broken out as Estimate phase view work starting with #4320)
- [ ] The Estimate phase shows a Parabol task at the top when a Parabol task is included in the scope

I’ll write issues as I verify the progress here.

**Comment by ackernaut** [comment]
Closed in favor of smaller issues. Voting UI will be worked out in #4320 and also added #4349 and #4350
